Implement on disk symbol file cache.

walac opened the pull request:
Add support for on disk symbol file cache. The cache is done through the pickle module. We maintain a list of the least recently used symbol files. When we reach the quota limit, we start to remove the symbol files until our disk usage fits inside the quota.

r? @vdjeric

Commit:
Move mkdir_p to the util module.

This function is useful in places other than symLogging.

vdjeric commented on sample-linux.ini:
```diff
@@ -10,6 +10,10 @@ maxCacheEntries = 2000000
 
+[DiskCache]
+cachePath = /tmp/snappy/cache
+cacheQuota = 1073741824
```
a gig seems like a lot. how many parsed & pickled xul.sym files is that? if it's > 10,000 files it might hurt disk lookup time?
in any case, make sure to test how well the server performs at the quota limit

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
+    for root, _, filenames in os.walk(path):
+      for filename in filenames:
+        filePath = os.path.join(root, filename)
+        stat = os.stat(filePath)
```
how long does stat'ing take on a hard-disk with 1GB worth of ~10KB files? That's roughly 100,000 stats
We could enforce the quota based on number of files instead. It's simpler, more performant, and "good enough"

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
+        filePath = os.path.join(root, filename)
+        stat = os.stat(filePath)
+        totalSize += stat.st_size
+        relativePath = os.path.relpath(filePath, self.cachePath)
```
isn't filename == relativePath?

walac replied:
`filename` is relative to `root`, `relativePath` is relative to `self.cachePath`.

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
+    return os.path.join(self.cachePath, pathSuffix)
+
+  def normPath(self, pathSuffix):
+    return os.path.normpath(pathSuffix)
```
Nit: might as well just call os.path.normpath directly

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
+      self.usedQuota -= fileSize
+
+  def absPath(self, pathSuffix):
+    r"Get the absolute path of the symbol file."
```
why is this string just hanging out? :)

walac replied:
This is a [doc string](https://www.python.org/dev/peps/pep-0257/), it is kind of standard way to document things on Python. It makes more sense on API's than applications, but not harmful, anyway.

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
+      path = self.absPath(path)
+      fileSize = os.path.getsize(path)
+      os.remove(path)
+      self.usedQuota -= fileSize
```
LogDebug which cache file you're removing, how large it was and current usedQuota

walac replied:
I fixed the algorithm to be based on file count instead of file size, so I think these extra info are less important now?

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
+    return os.path.normpath(pathSuffix)
+
+  def _walk(self, path):
+    r"Walk through the cache directory collecting files."
```
are these comments for the function?

walac replied:
Yep, as explained above.

vdjeric commented on symFileManager.py:
```diff
@@ -57,30 +158,30 @@ def GetLibSymbolMap(self, libName, breakpadId):
-
-      pathSuffix = os.path.join(libName, breakpadId, symFileName)
-      urlSuffix = "/".join([libName, breakpadId, symFileName])
+      libSymbolMap = self.diskCache.fetch(pathSuffix)
```
rest of the file is using capitals for the first letter of the function name, i.e. Fetch

vdjeric commented on symFileManager.py:
```diff
@@ -57,30 +158,30 @@ def GetLibSymbolMap(self, libName, breakpadId):
+            if libSymbolMap:
+              break
+
+        if libSymbolMap:
```
indentation is getting a bit much in this method, can you reduce it by splitting this into 2 methods and returning libSymbolMap as soon as it's found

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
     return self.entryCount
 
+# Manage in disk symbol cache
+class SymDiskCache(object):
```
might as well put this class in its own file

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
+
+    self.checkQuota()
+
+  def fetch(self, pathSuffix):
```
Use capital for first letter of function name

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
+    # crosses the size limit
+    self.usedQuota += os.path.getsize(path)
+
+  def use(self, pathSuffix):
```
call it filename instead of pathSuffix

vdjeric commented on symFileManager.py:
```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
+    r"Mark the cached sym file as recently used."
+    normPath = self.normPath(pathSuffix)
+    path = self.absPath(normPath)
+    if os.path.exists(path):
```
doing file I/O just to update an MRU list is bad

walac commented:
Comments addressed. Additionally, I replace the LRU list by a deque, so we have amortized O(1) performance for front elements removal.

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+      for filename in filenames:
+        filePath = os.path.join(root, filename)
+        stat = os.stat(filePath)
+        relativePath = os.path.relpath(filePath, self.cachePath)
```
we don't need to worry about paths being relative to root vs cachePath if we use a flat structure
are you planning to store the symbols in a nested directory structure?
if not, we don't need the extra relpath lines

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+    return symbolInfo
+
+  def Store(self, symbolInfo, filename):
+    r"""Store a symbol file in the cache.
```
I don't have a preference for docstrings vs comments above function signature, but I do want things to be consistent. So I guess convert all functions to docstrings? You can do that in another pull request

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+    r"Get the absolute path of the symbol file."
+    return os.path.join(self.cachePath, filename)
+
+  def _Walk(self, path):
```
ditto for consistency in using underscores for private methods

walac replied:
So, I am going to revert docstring to comments. If we are going to make code style refactory, I think we should be going with [PEP 8](https://www.python.org/dev/peps/pep-0008/), but I think it is only worth if we are interested in bring external contributors.

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+from symLogging import LogDebug
+from collections import deque
+
+# Manage in disk symbol cache
```
grammar: Singleton for managing the disk symbol cache
:)

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+    self.cachePath = cachePath
+    self.maxNumberOfFiles = maxNumberOfFiles
+
+    self.entryCount, fileList = self._Walk(cachePath)
```
you can just store fileList in the object so you don't have to keep hitting disk to check for non-existent files

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+        stat = os.stat(filePath)
+        relativePath = os.path.relpath(filePath, self.cachePath)
+        fileList.append((relativePath, stat.st_atime))
+      count += len(filenames)
```
can't the count be derived from len(fileList)?

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+    fileList.sort(cmp=lambda a,b: cmp(a[1], b[1]))
+
+    # filter out access times
+    self.lru = deque(map(lambda x: x[0], fileList))
```
nit: _Walk should return a sorted list without access times

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+    assert normPath not in self.lru, self.lru
+    self.lru.append(normPath)
+
+  def CheckCacheSize(self):
```
nit: if you're going to change state, don't call the function "Check", call it "EnforceCacheSize" or "EnforceQuota"

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+      path = self.absPath(path)
+      LogDebug("Evicting {} from cache.".format(path))
+      os.remove(path)
+      self.entryCount -= 1
```
entryCount is just len(self.lru), so there's no need to maintain a separate entryCount variable

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+
+  def CheckCacheSize(self):
+    r"Make sure we did not crosses the quota limit."
+    while self.entryCount > self.maxNumberOfFiles and self.lru:
```
you could simplify this:

evicted = self.lru[self.maxNumberOfFiles:]
self.lru = self.lru[:self.maxNumberOfFiles]

for file in evicted:
   # delete file

walac replied:
[deque](https://docs.python.org/2/library/collections.html#collections.deque) doesn't support slice operation :/

vdjeric replied:
ah ok, you can still simplify this. So instead:

for _ in range(self.maxNumberOfFiles, len(self.lru)):
  lru.pop()

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+      pickle.dump(symbolInfo, f, pickle.HIGHEST_PROTOCOL)
+
+    self.entryCount += 1
+    self.CheckCacheSize()
```
Think about what kind of eviction pattern this is going to create. Let's say the limit is 50 files. The server is going to cache 50 files and then get a request that caches 5 more files and will have to delete  the 5 oldest files. Then it will get another request and cache 7 new files and delete the 7 oldest again. And so on.

So after the cache reaches quota, every single request will have to pay the cost of evicting a whole bunch of files. That's not ideal.

How about when a request comes in that would cause to go over quota, you enqueue a new event to do quota maintenance. The maintenance event should delete say 40% of all files (based on age). You should also have a timer event to do the quota maintenance periodically (every 8 hours?).

vdjeric commented on sample-linux.ini:
```diff
@@ -10,6 +10,10 @@ maxCacheEntries = 2000000
 
+[DiskCache]
+cachePath = /tmp/snappy/cache
+maxCacheFiles = 50
```
50 is way too low, even a single request can have 50 libs. 
1. How big is one parsed+pickled xul.sym file?
2. Determine the maxCacheFiles limit based on using at most ~500MB of storage space filled with parsed+pickled xul.sym files.
3. How would the server have behaved in the situation when there were 60 libs needing to be stored (for one request) but the quota was 50?

walac replied:
`xul` and `libxul` seem to be outliers. They are 15MB size each, but most of the files are < 500KB.

vdjeric replied:
do you mean the raw .sym files are 15MB? How big are XULs after only the relevant lines are pickled?

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+    self.entryCount += 1
+    self.CheckCacheSize()
+
+  def Use(self, filename):
```
"use" is a very general name, how about UpdateStats?

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+
+  def Use(self, filename):
+    r"Mark the cached sym file as recently used."
+    normPath = os.path.normpath(filename)
```
why are paths being normalized and checked so much? we should just use one form of relative path internally. basically you just need to use the filename, don't worry about any directories or paths

walac replied:
I store the paths in the LRU, and make string comparison to locate them when I need to update the statiscs.

vdjeric replied:
why do we need to use paths at all if all the files are all in the same directory? we can just use the filenames

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+      self.lru.remove(normPath)
+    except ValueError:
+      pass
+    else:
```
this should be "finally", not else

walac replied:
`else` is right, I want to execute the `append` call only if the element was in `lru`.

vdjeric replied:
if the element isn't in the LRU, don't you want to append it to the front of the LRU?
Edit: I see now this is because you had a separate Use & Add

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+    else:
+      self.lru.append(normPath)
+
+  def Add(self, filename):
```
Do we need Add separate from Use? Whether you just stored the disk-cache file or if you just used an existing disk-cache file, its filename needs to move to the front of the LRU queue. So UpdateStats should do both Use and Store, no?

walac replied:
They are two operations because I want to keep the LRU consistent with memory cache, so when a memory cache entryis hit, I call `Use` to update the list, but it makes sense only if the element is there. `Add` is the opposite, add a element which is not there.

vdjeric replied:
this need goes away if you use a single LRU for both on-disk and in-memory caches, no?

vdjeric commented on symFileManager.py:
```diff
@@ -48,38 +50,46 @@ def GetLibSymbolMap(self, libName, breakpadId):
     if libName in self.sCache and breakpadId in self.sCache[libName]:
       libSymbolMap = self.sCache[libName][breakpadId]
       self.UpdateMruList(libName, breakpadId)
+      self.diskCache.Use(pathSuffix)
```
this GetLibSymbolMap function should be split into two functions, one for getting a libSymbolMap by any means, the other for maintaining the in-memory cache.

walac replied:
I didn't get this :/

vdjeric replied:
pseudo-code further below in another comment

vdjeric commented on symFileManager.py:
```diff
@@ -48,38 +50,46 @@ def GetLibSymbolMap(self, libName, breakpadId):
     libSymbolMap = None
     if libName in self.sCache and breakpadId in self.sCache[libName]:
       libSymbolMap = self.sCache[libName][breakpadId]
       self.UpdateMruList(libName, breakpadId)
```
so we really only need one LRU list. It will have the LRU info for the last let's say 1000 libraries and the disk-cache will have files for those 1000 libraries. The memory cache will only have in-memory version of the first 100 entries in the _same_ LRU. That one LRU gets updated based on incoming requests, i.e. it's just history of the last 1000 libraries used really.

walac replied:
So, thinking a bit about that, evicting gets problematic if we have only one LRU list for both caches, because, say, a memory eviction doesn't necessarily mean a disk eviction. What makes sense is to encapsulate all cache policy inside a class, say `SymbolCache`.

vdjeric replied:
Not sure i understand that argument, let's say the limit for in-memory cache is 3 and the limit for on-disk cache is 7.

So the LRU might look like this [a, b, c, d, e, f, g]. The elements in memory are lru[0:3] and the elements on disk are lru[:7]

If "g" get used, then the LRU looks like [g, a, b, c, d, e, f]. The memory cache is lru[:3] = [g, a, b]. 
That means "c" needs to be evicted from the memory-cache and "g" needs to be added in. The disk-cache is unchanged. It's the same behaviour as now.

if "h" gets added, then LRU looks like [h, g, a, b, c, d, e] which means "b" gets evicted from in-memory-cache, "f" gets evicted from disk cache, and "h" gets added to both the memory-cache and the disk-cache.

I don't think it's necessary to worry too much about data structure access/iteration efficiency during eviction. The LRU is going to have fewer than 1000 elements in the memory cache (representing 1000 files cached  in memory) and you won't need to iterate beyond that.

vdjeric commented on symFileManager.py:
```diff
@@ -220,3 +230,13 @@ def MaybeEvict(self, freeEntriesNeeded):
     #print "MRU after: " + str(self.sMruSymbols)
     LogDebug("Cache occupancy after MaybeEvict: " + str(self.sCacheCount) + "/" + str(maxCacheSize))
 
+  def _FetchSymbolsFromPaths(self, paths, pathSuffix, joinFunction, fetchFunction, cachePathSuffix):
```
this function is very general in a confusing way.. 
How about this?

GetSymbols:
1. symbols = FetchSymbols()
2. Update LRU stats

FetchSymbols:
1. check in-memory cache. if found, return
2. check on-disk cache. if found, insert into memory cache and return
3. if configured, check file paths
4. if configured, check URL paths
5. if symbol found, insert into memory cache & disk cache. if symbol not found, return failure

walac replied:
So, we have basically 4 symbol sources: memory cache, disk cache, local paths and remote paths. What I think would make the code easier to maintain is a "Fetcher" abstraction. I am going to work on something.

vdjeric commented on symbolicationWebService.py:
```diff
@@ -207,22 +224,16 @@ def ReadConfigFile():
-  if not set(["General", "Log"]).issubset(set(configParser.sections())):
-    logging.error("'General' and 'Log' sections are mandatory in the config file")
+  if not set(["General", "Log", "DiskCache"]).issubset(set(configParser.sections())):
+    logging.error("'General', 'Log' and 'DiskCache' sections are mandatory in the config file")
```
i don't know if we want a disk cache for local debugging. the user might change code, regenerate their local .sym files but forget to restart the Snappy server. that would result in an invalid disk-cache state

walac replied:
Hrm, I need to add code to handle when there is no cache configuration.

vdjeric commented on symDiskCache.py:
```diff
@@ -0,0 +1,97 @@
+    path = self.absPath(filename)
+    symbolInfo = None
+    if os.path.exists(path):
+      with open(path, 'r') as f:
```
From https://docs.python.org/2/library/pickle.html

"Note: Be sure to always open pickle files created with protocols >= 1 in binary mode."

vdjeric commented:
Basically, the design needs to be thought through better. Can you write up some short pseudocode for the cache behaviour in a Google Doc or something before making further code changes? These reviews will move faster once a design has been decided.

Commit:
Make the code pyflakes clean.

Small fixes caught by pyflakes.

Commit:
Implement on disk symbol file cache.

This commit adds support for on disk symbol file cache. The cache is
done through the pickle module. We maintain a list of the most recently
used symbol files. The MRU is shared between the memory and disk cache.
The first "maxEntries" entries in the MRU corresponds to the memory
cache.

We also refactor the symbol fetching code.

We apply some OOP to symbol fetching here. The concept of a symbol
fetcher object is a interface which will receive a lib and a breakpad id
and return the respective SymbolInfo object, or None if the SymbolInfo
object is not found.

There are currently two fetchers:
* PathFetcher: look for symbol files in local paths.
* URLFetcher: look for symbol files in a remote path.

The symbol file manager constructs a pipeline of fetchers, calling one
by one until either one of them returns a SymbolInfo or all fetchers
are called.

walac commented:
So, this new series of patches takes a lot of responsibility out of `SymFileManager`. First of all we now have the concept of a `Fetcher`: an abstraction which knows how to fetch a symbol file from a symbol source. It has two methods: `Fetch`, which takes a library and a breakpad id and returns a `SymbolInfo` , and `PostFetch`, which is a method that say "Hey, I got this SymbolInfo from some fetcher (might be you), here it is in case you need to know".

`SymFileManager` is now a [mediator](https://en.wikipedia.org/wiki/Mediator_pattern) among all the fetchers. It links the different fetchers and when `GetLibSymbolMap` is called, it goes on each fetcher (from the fastest to the slowest one) calling its `Fetch` method, until one returns the symbol info (or all fetchers are visited).

This new approach brings some improvements:
- `SymFileManager` got a lot simpler.
- The code is a bit more modular, if we want to add unit tests, the work just got a bit easier.
- Adding a new fetcher is relatively easier.
- The several pieces are decoupled.

There are two configurations for `DiskCacheFetcher` that need some discussion:
- The number of cache files (currently configured for 200 by default).
- The cache eviction rate (it evicts cache until it gets 50% empty).

Edit: I forgot to say about when the eviction happens. Currently it happens when the code detects the quota was exceeded. The side effect is that once a while a request will take a bit longer to finish.

Commit:
Handle ctrl-c on Windows.

select function on Windows does not return on SIGINT interrupt, so ctrl-c
doesn't stop the server. We add a timer to event loop to make select to
return and the server exit on ctrl-c.

Commit:
Use different log path for each process.

RotatingFileHandler has a bug on Windows which it cannot handle well
file roll over with multiple processes on Windows.

The solution here is to log data in different paths for each process.
The subprocess logs its data in the "subprocess" sub-directory.

vdjeric commented on symbolicationRequest.py:
```diff
@@ -14,11 +13,6 @@
 # for symbolication. Also prevents loops.
 MAX_FORWARDED_REQUESTS = 3
 
-class ModuleV3:
```
why remove this?

vdjeric commented on symbolicationRequest.py:
```diff
@@ -282,21 +277,20 @@ def Symbolicate(self, stackNum):
       module = self.combinedMemoryMap[moduleIndex]
 
-      if (module.libName, module.breakpadId) in missingSymFiles:
+      if module in missingSymFiles:
```
Ok, I think I see now. Did you test with V2 modules?

vdjeric commented on symbolicationWebService.py:
```diff
@@ -38,11 +37,7 @@
-  "mruSymbolStateFile": os.path.join(tempfile.gettempdir(), "snappy-mru-symbols.json"),
-  # Maximum number of symbol files to persist in the state file between runs.
-  "maxMRUSymbolsPersist": 10,
+  "maxCacheEntries": 10,
```
match the sample-linux.ini default?

vdjeric replied:
maxMemCacheFiles?

walac replied:
Fixed.

vdjeric commented on sample-windows.ini:
```diff
@@ -14,9 +14,11 @@ portNumber = 8000
-maxCacheEntries = 2000000
-mruSymbolStateFile = %(TEMP)s\snappy-mru-symbols.json
-maxMRUSymbolsPersist = 10
+maxCacheEntries = 400
```
The Windows symbol server is for profiling local builds where users might rebuild their libraries and we'll be symbolicating based on outdated symbols.

Can you add a switch so that symbols read from local disk do not get cached in either cache. Do this in a different pull request.

walac replied:
I changed the code for when `maxMemCacheFiles` and `maxDiskCacheFiles` are zero, cache is disabled. This is the default on `sample-windows.ini`.

vdjeric replied:
Can you make it so that files fetched remotely (i.e. URLFetcher fetching from S3) get cached in both disk caches), but not files read remotely (i.e. PathFetcher fetching from local disk)?
This is really an improvement, so leave it for a different pull request

vdjeric commented on symbolicationWebService.py:
```diff
@@ -50,7 +45,11 @@
-  ]
+  ],
+  # Symbol files cache path
+  "cachePath": os.path.join(tempfile.gettempdir(), 'snappy', 'cache'),
```
maybe call it diskCachePath?

walac replied:
Fixed.

vdjeric commented on symbolicationWebService.py:
```diff
@@ -50,7 +45,11 @@
+  # Symbol files cache path
+  "cachePath": os.path.join(tempfile.gettempdir(), 'snappy', 'cache'),
+  # Maximum number of cache files
+  "maxCacheFiles": 50
```
ditto.. maxDiskCacheFiles

walac replied:
Fixed.

vdjeric commented on symbolicationWebService.py:
```diff
@@ -212,17 +222,11 @@ def ReadConfigFile():
-        logging.error("Integer value expected for config option '" + option + "'")
-        return False
-    gOptions[option] = value
+  diskCacheSectionOptions = configParser.items("DiskCache")
```
where is "DiskCache" used?

vdjeric commented on sample-linux.ini:
```diff
@@ -6,9 +6,11 @@ portNumber = 8000
-maxCacheEntries = 2000000
-mruSymbolStateFile = /tmp/snappy-mru-symbols.json
-maxMRUSymbolsPersist = 10
+maxCacheEntries = 400
```
can you add a section for MemoryCache?

walac replied:
Added.

vdjeric commented on symbolicationWebService.py:
```diff
@@ -78,15 +77,13 @@ def initializeSubprocess(options):
-
-
 def processSymbolicationRequest(rawRequest, remoteIp):
   decodedRequest = json.loads(rawRequest)
```
what happens if this throws an exception?

walac replied:
The exception is caught, sent to the caller process, re-raised, handled by `SymbolHandler` and it returns an error to the client.

vdjeric commented on symFileManager.py:
```diff
@@ -1,232 +1,69 @@
-    if libName == "":
-      return None
+    self.MRU = self.diskCache.GetCacheEntries()
+    self.memoryCache.LoadCacheEntries(self.MRU, self.diskCache)
```
~~you don't really need to warm up the memory cache on startup~~
Oh I see you're doing it so that the MRU reflects the memory-cache state. You could also handle misses in memory cache lookups. I guess this is fine

vdjeric commented on symFileManager.py:
```diff
@@ -1,232 +1,69 @@
-    if libSymbolMap is None:
-      LogDebug("Need to fetch PDB file for " + libName + " " + breakpadId)
+  def GetLibSymbolMap(self, lib):
+    # Empty lib name means client couldn't associate frame with any lib
```
don't call GetLibSymbolMap if you know the lib name is bad

walac replied:
Fixed.

vdjeric commented on symCache.py:
```diff
@@ -0,0 +1,111 @@
+  def __init__(self, options):
+    super(MemoryCache, self).__init__(options)
+    self.sCache = {}
+    self.sCacheSize = 0
```
is sCacheSize used?

walac replied:
Artifact from code clean-up, removed.

vdjeric commented on symCache.py:
```diff
@@ -0,0 +1,111 @@
+      self.sCache[lib] = symbols[lib]
+
+  def Get(self, lib):
+    return self.sCache[lib]
```
Maybe .get() here instead so it doesn't throw exceptions?
That would also make it consistent with DiskCache.Get()

Update: I see now that you rely on this method throwing an exception. Returning None is better than relying on exceptions

walac replied:
By code logic, a `KeyError` here is a bug, anyway I changed it to use `get`.

vdjeric commented on symCache.py:
```diff
@@ -0,0 +1,111 @@
+            self.sOptions["cachePath"],
+            libName,
+            breakpadId,
+            GetSymbolFileName(libName))
```
use a flat directory structure, e.g. cacheDir/xul.pdb-ABCDEF0A1.pickle
if you create nested directories, eviction will only delete files and not delete empty directories, which means the server will forever accumulate empty directories until they're manually cleaned out
You also do less I/O when storing files.
You don't have to use a hyphen to separate the parts, use one of the characters that is not allowed by gLibNameRE

walac replied:
Fixed.

vdjeric commented on symCache.py:
```diff
@@ -0,0 +1,111 @@
+class DiskCache(Cache):
+  def __init__(self, options):
+    super(DiskCache, self).__init__(options)
+    self.sOptions = options
```
nit: you don't really need to store all the options, you only use maxCacheFiles and cachePath

walac replied:
Fixed.

vdjeric commented on symFileManager.py:
```diff
@@ -1,232 +1,69 @@
+        cache = self.diskCache
+      LogDebug("Loading [{}] [{}] from {}".format(lib[0], lib[1], cache.__class__))
+      libSymbolMap = cache.Get(lib)
+    except ValueError:
```
doesn't DiskCache.Get() return None if it can't find the file on disk? that means ValueError wouldn't get thrown, right?

walac replied:
Good point, fixed.

vdjeric commented on symFetcher.py:
```diff
@@ -0,0 +1,90 @@
+  # function is responsible for some post processing,
+  # like symbol caching
+  def PostFetch(self, libSymbolMap, libName, breakpadId):
+    pass
```
I don't think PostFetch is needed anymore?

walac replied:
You are right, removed.

vdjeric commented on symFetcher.py:
```diff
@@ -0,0 +1,90 @@
+          return None
+        headers = request.info()
+        contentEncoding = headers.get("Content-Encoding", "").lower()
+        if contentEncoding in ("gzip", "x-gzip", "deflate"):
```
what happens on the else branch? can ParseSymbolFile() read from the object returned by urllib2.urlopen (i.e. "request") directly?

walac replied:
Yes.

vdjeric commented on sample-linux.ini:
```diff
@@ -5,10 +5,12 @@ portNumber = 8000
+
+[DiskCache]
+diskCachePath = /tmp/snappy/cache
+maxDiskCacheFiles = 1500
```
Assuming most files are like xul.sym, and averege xul.sym pickled size is 15mb, this is about 22.5GB.
Looking at the centos machine, I see:

Filesystem      Size  Used Avail Use% Mounted on
/dev/xvda1      8.6G  2.8G  5.8G  33% /

The maxDiskCacheFiles will have to be smaller until there's more storage available. So 5.8GB / 15MB per file ~= 387. Let's set maxDiskCacheFiles to 300 files

walac replied:
Going with more realistic statistics, I created a [small script](https://gist.github.com/walac/fa72f6b15327e7b332a9) to calculate the average file size from production logs. The results showed less than 800KB per file. Being a bit more conservative we can assume 1MB per file, so 1500 files is still a reasonable estimate.

vdjeric replied:
Ok let's try it out

vdjeric commented:
I wrote this on an outdated diff, so I'm not sure if it's visible:

Can you make it so that files fetched remotely (i.e. URLFetcher fetching from S3) get cached in both disk caches), but not files read remotely (i.e. PathFetcher fetching from local disk)?
This is really an improvement, so leave it for a different pull request

vdjeric commented:
This code is almost ready for merge. Did you test it end-to-end on Windows & Linux?

vdjeric commented:
Btw, you can remove the V2 support, I confirmed nothing uses it anymore. Do this in a separate pull request.

walac commented:
> Can you make it so that files fetched remotely (i.e. URLFetcher fetching from S3) get cached in both disk caches), but not files read remotely (i.e. PathFetcher fetching from local disk)?
> This is really an improvement, so leave it for a different pull request

I am not sure if I understood, do you want to cache files only when fetched from S3?

walac commented:
> This code is almost ready for merge. Did you test it end-to-end on Windows & Linux?

Yes.

walac commented:
> Btw, you can remove the V2 support, I confirmed nothing uses it anymore. Do this in a separate pull request.

Ok.

vdjeric commented:
> I am not sure if I understood, do you want to cache files only when fetched from S3?

I was saying that when a developer starts the server with sample-windows.ini (i.e. on their local machine), I'd only like to store files fetched from S3 in the memory cache or disk cache. I don't want the server to cache any files it fetches from local disk in any of the caches.

If the user starts the serverwith sample-linux.ini, I want the current behavior you have in this pull request.

vdjeric commented:
I deployed it on centos, it's looking good
